### ISSUE TITLE
Context menu is slow to show up

### ISSUE THREAD

**Issue by kheffah** [DigitalSlideArchive/HistomicsTK#533]
I don't know why, but there is always significant delay when I right click to get the context menu. It may be querying the database and looping though all annotations or something, but it's definitely much slower to respond than would be needed in a high-throughput project. Not a high priority issue, but would definitely be great if we can pinpoint and eliminate the cause for the delay. Thanks!

**Comment by manthey** [comment]
@kheffah Is this still an issue?  I can't replicate it on current master, even on annotations with 10,000+ elements.

**Comment by kheffah** [comment]
@manthey Yes, it is still an issue unfortunately. There is a ~3-4 second delay when I right click on any annotation. This issue exists in the slide I send you before where the slider transparency was problematic. You may use the same annotation set to test it. 3-4 seconds delay may seem small, but when tens of even 100+ corrections are needed, this becomes an unfavorable penalty. Thanks!

**Comment by manthey** [comment]
I'm not seeing it on that slide and annotations on current master, but it could just be I have fast hardware.  Some of the changes in the last week or two may have improved this, but I doubt it would be substantial.  Do you know if you are using the latest dockers?

Looking at the code, I think we are recomputing things we don't need to.

**Comment by kheffah** [comment]
@manthey I'm using the docker version just before the last couple of PRs (about downloading annotation styles), not sure if that's the latest one. It would be great if any duplicate calculations are minimized since the context menu us a very heavily used feature. Thanks!

**Comment by manthey** [comment]
This should be improved with https://github.com/OpenGeoscience/geojs/pull/987.  That got merged into master yesterday.  There is always more room for improvement, but hopefully that will make it perform well enough in your environment.

**Comment by manthey** [comment]
I have a series of PRs which should help this quite a bit:
https://github.com/OpenGeoscience/geojs/pull/994
https://github.com/girder/large_image/pull/345
#633 

Basically, the slow down occurred when there were lots of annotations, not necessarily when there were lots of elements.

We were rerendering the annotation selection panel multiple times (up to once per annotation).  The time to construct and deconstruct backbone tooltips is weirdly high.  Resetting empty backbone collections triggers a series of events which are unnecessary.  We would rerender each time an annotation changed; it is better to batch them together on animation frames.  Lastly, we were doing more work in webgl when points changed than was necessary.  Because this last issue happens in conjunction with the GPU, the web performance analyzers shows the time in a function that has virtually no relevance to the actual issue, which made it a bit trickier to hunt down.

**Comment by kheffah** [comment]
@manthey I see, glad to see this was tracked down. Thank you so much @manthey !

**Comment by manthey** [comment]
I'm closing this since all of the various PRs have made it in.  Feel free to open a new issue as needed.

**Comment by kheffah** [comment]
I am re-opening this because it's causing problems again. Not sure what is wrong, but when the number of elements is large the context menu is very slow to show up (I am using the later docker version). This time, I think I have a helpful clue: the "label" is very fast, unlike the context menu. In a previous issue (issue #632), some logic was introduced to allow the label to only show the innermost label for an element. So the exact **same** kind of lookup should be used in the context menu should be used, if possible.
